File: docs/manual/scala/guide/services/code/ServiceImplementation.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl.services

package helloservice {
  //#hello-service-impl
  import com.lightbend.lagom.scaladsl.api.ServiceCall
  import scala.concurrent.Future

  class HelloServiceImpl extends HelloService {
    override def sayHello = ServiceCall { name =>
      Future.successful(s"Hello $name!")
    }
  }
  //#hello-service-impl
}

package lagomapplication {

  import helloservice._

  //#lagom-application
  import com.lightbend.lagom.scaladsl.server._
  import play.api.libs.ws.ahc.AhcWSComponents
  import com.softwaremill.macwire._

  abstract class HelloApplication(context: LagomApplicationContext)
      extends LagomApplication(context)
      with AhcWSComponents {

    override lazy val lagomServer = serverFor[HelloService](wire[HelloServiceImpl])
  }
  //#lagom-application
}

package lagomloader {

  import lagomapplication._
  import helloservice._

  //#lagom-loader
  import com.lightbend.lagom.scaladsl.server._
  import com.lightbend.lagom.scaladsl.api.ServiceLocator
  import com.lightbend.lagom.scaladsl.devmode.LagomDevModeComponents

  class HelloApplicationLoader extends LagomApplicationLoader {

    override def loadDevMode(context: LagomApplicationContext) =
      new HelloApplication(context) with LagomDevModeComponents

    override def load(context: LagomApplicationContext) =
      new HelloApplication(context) {
        override def serviceLocator = ServiceLocator.NoServiceLocator
      }

    override def describeService = Some(readDescriptor[HelloService])
  }
  //#lagom-loader
}

package callstream {

  import akka.NotUsed
  import com.lightbend.lagom.scaladsl.api.ServiceCall

  import scala.concurrent.Future

  class CallStreamImpl extends CallStream {
    //#tick-service-call
    import scala.concurrent.duration._
    import akka.stream.scaladsl.Source

    override def tick(intervalMs: Int) = ServiceCall { tickMessage =>
      Future.successful(
        Source
          .tick(
            intervalMs.milliseconds,
            intervalMs.milliseconds,
            tickMessage
          )
          .mapMaterializedValue(_ => NotUsed)
      )
    }
    //#tick-service-call
  }
}

package hellostream {

  import com.lightbend.lagom.scaladsl.api.ServiceCall
  import scala.concurrent.Future

  class HelloStreamImpl extends HelloStream {
    //#hello-service-call
    override def sayHello = ServiceCall { names =>
      Future.successful(names.map(name => s"Hello $name!"))
    }
    //#hello-service-call
  }
}

package serverservicecall {

  import scala.concurrent.Future

  class UsingServerServiceCall extends helloservice.HelloService {
    //#server-service-call
    import com.lightbend.lagom.scaladsl.server.ServerServiceCall
    import com.lightbend.lagom.scaladsl.api.transport.ResponseHeader

    override def sayHello = ServerServiceCall { (requestHeader, name) =>
      val user = requestHeader.principal
        .map(_.getName)
        .getOrElse("No one")
      val response = s"$user wants to say hello to $name"

      val responseHeader = ResponseHeader.Ok
        .withHeader("Server", "Hello service")

      Future.successful((responseHeader, response))
    }
    //#server-service-call
  }
}

package servicecallcomposition {

  import scala.concurrent.ExecutionContext
  import scala.concurrent.Future
  import com.lightbend.lagom.scaladsl.server.ServerServiceCall

  object Logging {
    //#logging-service-call
    def logged[Request, Response](serviceCall: ServerServiceCall[Request, Response]) =
      ServerServiceCall.compose { requestHeader =>
        println(s"Received ${requestHeader.method} ${requestHeader.uri}")
        serviceCall
      }
    //#logging-service-call
  }

  class LoggedHelloService extends helloservice.HelloService {
    import Logging.logged

    //#logged-hello-service
    override def sayHello =
      logged(ServerServiceCall { name =>
        Future.successful(s"Hello $name!")
      })
    //#logged-hello-service
  }

  trait User

  //#user-storage
  trait UserStorage {
    def lookupUser(username: String): Future[Option[User]]
  }
  //#user-storage

  object Authentication {
    val userStorage: UserStorage      = null
    implicit val ec: ExecutionContext = null

    //#auth-service-call
    import com.lightbend.lagom.scaladsl.api.transport.Forbidden

    def authenticated[Request, Response](
        serviceCall: User => ServerServiceCall[Request, Response]
    ) = ServerServiceCall.composeAsync { requestHeader =>
      // First lookup user
      val userLookup = requestHeader.principal
        .map(principal => userStorage.lookupUser(principal.getName))
        .getOrElse(Future.successful(None))

      // Then, if it exists, apply it to the service call
      userLookup.map {
        case Some(user) => serviceCall(user)
        case None       => throw Forbidden("User must be authenticated to access this service call")
      }
    }
    //#auth-service-call
  }

  class AuthHelloService extends helloservice.HelloService {
    import Authentication.authenticated
    //#auth-hello-service
    override def sayHello = authenticated { user =>
      ServerServiceCall { name =>
        Future.successful(s"$user is saying hello to $name")
      }
    }
    //#auth-hello-service
  }

  object Filter {
    import Logging.logged
    import Authentication.authenticated

    //#compose-service-call
    def filter[Request, Response](
        serviceCall: User => ServerServiceCall[Request, Response]
    ) = logged(authenticated(serviceCall))
    //#compose-service-call
  }

  class FilterHelloService extends helloservice.HelloService {
    import Filter.filter
    //#filter-hello-service
    override def sayHello = filter { user =>
      ServerServiceCall { name =>
        Future.successful(s"$user is saying hello to $name")
      }
    }
    //#filter-hello-service
  }

}

File: docs/manual/java/guide/services/code/docs/services/ServiceImplementation.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.services;

import akka.japi.Pair;
import akka.stream.javadsl.Source;
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.transport.Forbidden;
import com.lightbend.lagom.javadsl.api.transport.ResponseHeader;
import com.lightbend.lagom.javadsl.server.HeaderServiceCall;
import com.lightbend.lagom.javadsl.server.ServerServiceCall;
import scala.concurrent.duration.FiniteDuration;

import java.security.Principal;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.time.Duration;

import static java.util.concurrent.CompletableFuture.completedFuture;

public class ServiceImplementation {

  public static class HeaderServiceCallLambda implements HelloService {
    // #header-service-call-lambda
    public HeaderServiceCall<String, String> sayHello() {
      return (requestHeader, name) -> {
        String user = requestHeader.principal().map(Principal::getName).orElse("No one");
        String response = user + " wants to say hello to " + name;

        ResponseHeader responseHeader = ResponseHeader.OK.withHeader("Server", "Hello service");

        return completedFuture(Pair.create(responseHeader, response));
      };
    }
    // #header-service-call-lambda
  }

  public static class HeaderServiceCallOfLambda implements HelloService {
    // #header-service-call-of-lambda
    public ServerServiceCall<String, String> sayHello() {
      return HeaderServiceCall.of(
          (requestHeader, name) -> {
            String user = requestHeader.principal().map(Principal::getName).orElse("No one");
            String response = user + " wants to say hello to " + name;

            ResponseHeader responseHeader = ResponseHeader.OK.withHeader("Server", "Hello service");

            return completedFuture(Pair.create(responseHeader, response));
          });
    }
    // #header-service-call-of-lambda
  }

  public static class TickServiceCall implements FirstDescriptor.CallStream {
    // #tick-service-call
    public ServerServiceCall<String, Source<String, ?>> tick(int intervalMs) {
      return tickMessage -> {
        Duration interval = Duration.ofMillis(intervalMs);
        return completedFuture(Source.tick(interval, interval, tickMessage));
      };
    }
    // #tick-service-call
  }

  public static class HelloServiceCall implements FirstDescriptor.HelloStream {
    // #hello-service-call
    public ServerServiceCall<Source<String, ?>, Source<String, ?>> sayHello() {
      return names -> completedFuture(names.map(name -> "Hello " + name));
    }
    // #hello-service-call
  }

  public static class ServiceCallComposition {

    // #logging-service-call
    public <Request, Response> ServerServiceCall<Request, Response> logged(
        ServerServiceCall<Request, Response> serviceCall) {
      return HeaderServiceCall.compose(
          requestHeader -> {
            System.out.println("Received " + requestHeader.method() + " " + requestHeader.uri());
            return serviceCall;
          });
    }
    // #logging-service-call

    public class LoggedHelloService implements HelloService {
      // #logged-hello-service
      public ServerServiceCall<String, String> sayHello() {
        return logged(name -> completedFuture("Hello " + name));
      }
      // #logged-hello-service
    }

    public static class User {}

    // #user-storage
    interface UserStorage {
      CompletionStage<Optional<User>> lookupUser(String username);
    }
    // #user-storage

    private UserStorage userStorage;

    // #auth-service-call
    public <Request, Response> ServerServiceCall<Request, Response> authenticated(
        Function<User, ServerServiceCall<Request, Response>> serviceCall) {
      return HeaderServiceCall.composeAsync(
          requestHeader -> {

            // First lookup user
            CompletionStage<Optional<User>> userLookup =
                requestHeader
                    .principal()
                    .map(principal -> userStorage.lookupUser(principal.getName()))
                    .orElse(completedFuture(Optional.empty()));

            // Then, if it exists, apply it to the service call
            return userLookup.thenApply(
                maybeUser -> {
                  if (maybeUser.isPresent()) {
                    return serviceCall.apply(maybeUser.get());
                  } else {
                    throw new Forbidden("User must be authenticated to access this service call");
                  }
                });
          });
    }
    // #auth-service-call

    public class AuthHelloService implements HelloService {
      // #auth-hello-service
      public ServerServiceCall<String, String> sayHello() {
        return authenticated(user -> name -> completedFuture("Hello " + user));
      }
      // #auth-hello-service
    }

    // #compose-service-call
    public <Request, Response> ServerServiceCall<Request, Response> filter(
        Function<User, ServerServiceCall<Request, Response>> serviceCall) {
      return logged(authenticated(serviceCall));
    }
    // #compose-service-call

    public class FilterHelloService implements HelloService {
      // #filter-hello-service
      public ServerServiceCall<String, String> sayHello() {
        return filter(user -> name -> completedFuture("Hello " + user));
      }
      // #filter-hello-service
    }
  }
}
